Cluster Management

cmeiklejohn opened the pull request:
Our awesome new cluster management page.
- [x] Add node.
- [x] Implement ring not ready state.
- [x] Implement legacy state.
- [x] Poll for updates to the plan.
- [x] API for updating the plan.
- [x] Wire up clear plan button to delete action.
- [x] Wire up commit plan action.
- [x] No plan state
- [x] Loading indicators
- [x] Error state indicators
- [x] Verify CSRF token protection works
- [x] Figure out bug when there is no plan and it doesn't update
- [x] When removing one node from two, it shows claim incorrectly.
- [x] Come up with custom error states (last, nice to have)
- [x] Ensure post works with no body.
- [x] Verify button is clicked.
- [x] State for no replacement nodes
- [x] Front page fixed width font
- [x] spans should be labels in action and for comit
- [x] Colorized arrows
- [x] Close notifications
- [x] Add warning for changes to node hosting riak control.
- [x] Disable actions on nodes after they've been staged; re-enable if plan is cleared

jgnewman committed:
more css, more responsiveness, new column for stage buttons, new area for planned cluster

jgnewman committed:
basic markup structure is there

cmeiklejohn committed:
Add new resource to return the cluster and plan.

cmeiklejohn committed:
Return the plan to the client.

cmeiklejohn committed:
Add resource to clear the plan.

cmeiklejohn committed:
Remove riak_core_ring_events watcher.

cmeiklejohn committed:
Move plan clear into the gen_server.

cmeiklejohn committed:
Move plan retrieval into the gen_server.

jgnewman committed:
new markup for ember on cluster page

jgnewman committed:
first ui pass; current cluster view

cmeiklejohn committed:
Add additional memory fields in.

cmeiklejohn committed:
Fix spelling error.

cmeiklejohn committed:
Using the future ring, compute claim and merge.

cmeiklejohn committed:
Fix typing errors, and use the new plan.

cmeiklejohn committed:
Refactor to make code cleaner.

cmeiklejohn committed:
Update to use correct API.

cmeiklejohn committed:
Fix epic fail.

cmeiklejohn committed:
Merge branch 'csm-add-planning-api' into jgn-new-ui

cmeiklejohn committed:
Start integrating with backend, setup requests.

cmeiklejohn committed:
Fix failure when there is no plan.

cmeiklejohn committed:
Integrate with backend.

cmeiklejohn committed:
Code cleanup.

cmeiklejohn committed:
Dialyzer fixes.

cmeiklejohn committed:
First pass at displaying the cluster plan.

cmeiklejohn committed:
Normalize values for now until core is patched.

cmeiklejohn committed:
Don't build nodes if they aren't an array.

cmeiklejohn committed:
Remove legacy JS.

cmeiklejohn committed:
Remove deprecated cluster controls.

cmeiklejohn committed:
Fix rendering of action.

cmeiklejohn committed:
Rename field.

jgnewman committed:
we now have a clear plan button and we only show the plan when the ring is ready, the controller is done loading, and we have at least one item in the plan

cmeiklejohn committed:
Remove plan details and add legacy state.

cmeiklejohn committed:
Rid the jshint errors.

jgnewman committed:
added some markup for legacy state

cmeiklejohn committed:
Add object for modeling the clusters.

cmeiklejohn committed:
Integrate new clusters object.

cmeiklejohn committed:
Update to use new memory efficient polling.

cmeiklejohn committed:
Updated staged cluster.

cmeiklejohn committed:
Fix rebase problem.

cmeiklejohn committed:
Implement ring_not_ready and legacy.

cmeiklejohn commented:
Just wired up ring_not_ready and legacy and got the polling working for every 1.5 seconds.

cmeiklejohn commented:
Wired up an efficient poll every 1.5 seconds which doesn't recreate Ember arrays or objects when things change.

cmeiklejohn committed:
Handle the no-plan state.

cmeiklejohn committed:
Add cluster plan clear button.

cmeiklejohn committed:
Fix bad $.ajax call.

cmeiklejohn committed:
Fix delete resource.

cmeiklejohn committed:
Add functional clear plan.

cmeiklejohn committed:
Remove useless comment.

seancribbs commented on priv/admin/js/cluster.js:
```diff
@@ -3,11 +3,215 @@ minispade.register('cluster', function() {
+     * @returns {void}
+     */
+    init: function() {
+      this.set('content',
```
Shouldn't this be set by the router?

cmeiklejohn replied:
Yes, it should.  Will fix.

cmeiklejohn replied:
Fixed in [f58590d].

seancribbs commented on priv/admin/js/cluster.js:
```diff
@@ -21,4 +225,283 @@ minispade.register('cluster', function() {
+
+      var self = this;
+
+      $.ajax({
```
I'd move handling of the action to the controller, or perhaps instead bind this with the action helper in the template, e.g.

``` html
<a class="gui-rect-button gui-text-bold" {{action clearPlan}}>Clear</a>
```

cmeiklejohn replied:
Resolved in [24d94be].

seancribbs commented on priv/admin/js/cluster.js:
```diff
@@ -21,4 +225,283 @@ minispade.register('cluster', function() {
+      var prop = this.get('parentView.expanded');
+
+      if (prop) {
+        this.get('parentView').set('expanded', false);
```
Could this be done with an `{{#if}}` in the template instead?

cmeiklejohn replied:
Leaving this for now as we need to build out this functionality as this feature grows with the ability to stage changes.

seancribbs commented on priv/admin/js/cluster.js:
```diff
@@ -21,4 +225,283 @@ minispade.register('cluster', function() {
+    /* Bindings from the model */
+
+    templateName:       'current_cluster_item',
+    nameBinding:        'content.name',
```
This smacks of scoping issues. Is `{{#with view}}` being used in the template instead of directly referring to properties of the controller/proxy?

seancribbs replied:
Also, I'd note that bindings are more expensive than computed properties because they aren't one-way.

cmeiklejohn replied:
So, you're saying you want it to be [one way](http://www.youtube.com/watch?v=409Pjtq7jzY)?

reiddraper replied:
You want it to be one way. [But I can't go for that](http://www.youtube.com/watch?v=ccenFp_3kq8).

seancribbs commented on priv/admin/js/cluster.js:
```diff
@@ -21,4 +225,283 @@ minispade.register('cluster', function() {
+   *
+   * Collection view for showing the current cluster.
+   */
+  RiakControl.CurrentClusterView = Ember.CollectionView.extend(
```
I'd prefer `{{#each}}` instead of a reified view class.

cmeiklejohn replied:
Since bound handlebars helpers are not available, this is the cleanest way to keep the formatting code on the view instance for each individual row.  However, I agree that the other approach is more appropriate in later versions of Ember.

seancribbs commented on priv/admin/js/cluster.js:
```diff
@@ -21,4 +225,283 @@ minispade.register('cluster', function() {
+     *
+     * @returns {boolean}
+     */
+    isReplaced: function() {
```
Seems like controller/model functionality.

cmeiklejohn replied:
Moved to model, thanks. [9a5d4aa]

seancribbs commented on priv/admin/js/cluster.js:
```diff
@@ -21,4 +225,283 @@ minispade.register('cluster', function() {
+     *
+     * @returns {string}
+     */
+     indicatorLights: function() {
```
Could this be a mixin? Seems like it's used in more than one place.

cmeiklejohn replied:
We use it somewhere else, but the color scheme is different there, so we can't extract it as is.

cmeiklejohn committed:
Fix initialize.

cmeiklejohn committed:
Move to model.

cmeiklejohn committed:
Switch Button View to action.

cmeiklejohn committed:
Add jsdoc.

cmeiklejohn committed:
Add backend for staging cluster changes.

cmeiklejohn committed:
Handle POSTs to update the cluster plan.

cmeiklejohn committed:
POST commits the plan.

cmeiklejohn committed:
Switch addNode to action, add stage handler.

cmeiklejohn committed:
Wire up join action in the backend.

cmeiklejohn commented:
Completed backend for staging changes, joining nodes, and committing the plan.

jgnewman committed:
new actions panel including new radio buttons and functionality for them

jgnewman committed:
fixed the bound name attribute on radio buttons.  had a test word in there before

cmeiklejohn committed:
Move action commands to the gen_server.

cmeiklejohn committed:
Merge branch 'jgn-new-ui' of github.com:basho/riak_control into jgn-new-ui

jgnewman committed:
we now have a display message for an empty plan

jgnewman committed:
Merge branch 'jgn-new-ui' of github.com:basho/riak_control into jgn-new-ui

cmeiklejohn committed:
Remove stray invalid preventDefault.

cmeiklejohn committed:
Add CSRF prevention methods.

cmeiklejohn committed:
Make add node button a view.

cmeiklejohn committed:
Force 409 return code when PUT request fails.

cmeiklejohn committed:
Automatically add CSRF header to all AJAX requests.

cmeiklejohn committed:
Complete add node feature.

cmeiklejohn committed:
Add CSRF protection.

cmeiklejohn committed:
Merge branch 'jgn-new-ui' of github.com:basho/riak_control into jgn-new-ui

Conflicts:
	priv/admin/js/generated/templates.js

cmeiklejohn commented:
Added CSRF protection.

cmeiklejohn commented:
Completed add node.

cmeiklejohn committed:
Remove admin_node.

jgnewman committed:
some preliminary error display stuff

cmeiklejohn committed:
Completely remove node routes.

cmeiklejohn committed:
Add stage handler.

cmeiklejohn committed:
Wire up replacement node select box.

cmeiklejohn committed:
Wire up stage action.

cmeiklejohn committed:
Make error more generic.

cmeiklejohn committed:
Bump templates.

cmeiklejohn committed:
Fix claim percentages.

If a node doesn't appear in the future claim, it's claim is 0.

cmeiklejohn commented:
Pushed an update for incorrect claim computation.

cmeiklejohn committed:
Fix UI glitches with state updating.

cmeiklejohn committed:
Formatting; wire up commit button.

cmeiklejohn committed:
Ensure the plan is confirmed.

cmeiklejohn committed:
Ensure that we can handle no body.

cmeiklejohn commented:
Implemented verify body, implemented handling for no body.  @jgnewman can you make the confirm text a label so we do not need to click the box directly?

cmeiklejohn committed:
Return better error messages.

cmeiklejohn commented:
Just pushed a minor update to get better error messages when staging changes.

cmeiklejohn committed:
Fix typing error.

jgnewman committed:
state for no replacement nodes; front page fixed width font; spans are now labels in action and commit

jgnewman commented:
@cmeiklejohn Hey, when I try to add a bad node, the error message doesn't come up.  Looks like in cluster.js you're calling JSON.parse but the server's giving us HTML instead of JSON.

cmeiklejohn commented:
I've got a fix for this and will push it later this afternoon.

Chris

Sent from my iPhone

On Apr 4, 2013, at 15:05, John Newman notifications@github.com wrote:

> @cmeiklejohn Hey, when I try to add a bad node, the error message doesn't come up. Looks like in cluster.js you're calling JSON.parse but the server's giving us HTML instead of JSON.
> 
> —
> Reply to this email directly or view it on GitHub.

cmeiklejohn committed:
Add nodedown error.

jgnewman committed:
clicking the x now closes errors

cmeiklejohn committed:
Add pending_pct back.

cmeiklejohn committed:
Use the replacement node name.

cmeiklejohn commented on src/admin_routes.erl:
```diff
@@ -19,27 +19,24 @@
 %% routes that query/act on the cluster
-cluster_route (Route) -> admin_route(["cluster"|Route]).
-cluster_route () -> admin_route(["cluster"]).
+cluster_route(Route) -> admin_route(["cluster"|Route]).
```
I think this can be removed entirely.

cmeiklejohn commented on priv/admin/js/legacy/gui.js:
```diff
@@ -58,19 +62,64 @@ $(document).ready(function() {
+      }
+  });
+
+  /*
```
@jgnewman We shouldn't have commented code committed.  Can we nuke this?

jgnewman replied:
Yeah that's been the plan all along I thought.  We were going to get the cluster page up and working then delete all the stuff we were sure we didn't need anymore in legacy.js.

cmeiklejohn commented on priv/admin/js/legacy/gui.js:
```diff
@@ -8,19 +8,21 @@
   $.riakControl = $.riakControl || {};
 
-  /* ALLOWS YOU TO HIT ENTER IN THE ADD-NODE FIELD WITHOUT MAKING IT A FORM */
+  /*
```
@jgnewman More commented code.

jgnewman committed:
warning message for actions on the node hosting riak control

cmeiklejohn committed:
Ensure we populate pending_pct.

seancribbs commented on src/admin_cluster.erl:
```diff
@@ -0,0 +1,324 @@
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
```
A nice `%% @doc` with what this resource is/does might be nice. Same goes for the rest of them!

cmeiklejohn replied:
Resolved in [85ab285bd15f60b01fa68a13d128b84919770a10].

cmeiklejohn committed:
Fix typo.

cmeiklejohn committed:
Kill console.log.

seancribbs commented on src/admin_cluster.erl:
```diff
@@ -0,0 +1,324 @@
+
+%% @doc Given a struct/proplist that we've received via JSON,
+%% recursively turn the keys into atoms from binaries.
+atomize({struct, L}) ->
```
This function makes you vulnerable to an atom injection attack. Use binary_to_existing_atom(I, utf8) instead. If necessary, define the needed atoms somewhere in this module.

cmeiklejohn replied:
Resolved by using binary_to_existing_atom.

seancribbs commented on src/admin_cluster.erl:
```diff
@@ -0,0 +1,324 @@
+    Result = proplists:get_value(Key, List),
+    case is_binary(Result) of
+        true ->
+            binary_to_atom(Result, utf8);
```
Again, use `binary_to_existing_atom/2` if it is at all possible the binary will come from the user.

seancribbs commented on src/admin_cluster.erl:
```diff
@@ -0,0 +1,324 @@
+    Result = proplists:get_value(Key, List, Default),
+    case is_binary(Result) of
+        true ->
+            binary_to_atom(Result, utf8);
```
... DITTO

seancribbs commented on src/admin_cluster.erl:
```diff
@@ -0,0 +1,324 @@
+        self_join ->
+            "Node can not be joined to itself."
+    end,
+    list_to_binary(ErrorMessage).
```
Is it necessary to use list_to_binary here? The webmachine response can be an iolist.

seancribbs commented on src/admin_cluster.erl:
```diff
@@ -0,0 +1,324 @@
+-spec stage_individual_change({struct, term()}, boolean()) -> boolean().
+stage_individual_change({struct, Change}, Exit) ->
+    Node = atomized_get_value(node, Change),
+    Action = atomized_get_value(action, Change),
```
Another thought, you could avoid atomizing everything here by using `kvc`, which will traverse mochijson2-style structs for you.

cmeiklejohn replied:
Switched to binary_to_existing_atom.

seancribbs commented on src/admin_routes.erl:
```diff
@@ -19,27 +19,24 @@
+vnode_route(Route) -> admin_route(["vnode",partition|Route]).
+vnode_route() -> admin_route(["vnode",partition]).
 
 %% new restful routes
```
Routes ain't restful, bro. :trollface:

cmeiklejohn committed:
Change export, remove unused route.  Change to iolist.

cmeiklejohn committed:
Fix spec.

jgnewman committed:
colored lights; disabled actions on current cluster nodes after they have been staged

jgnewman committed:
fixed merge conflict

jgnewman committed:
recompiled hbs after merge conflict

cmeiklejohn committed:
Add documentation.

cmeiklejohn committed:
Don't build_plt against deps as it crashes the dialyzer.

seancribbs commented:
Just to record it for posterity, @cmeiklejohn mentioned the only case of where atoms MUST be created is nodenames, however a quick check against `epmd` can verify that the new nodename is ok, should the atom not already exist.

``` text
$ erl -name foo@127.0.0.1
# snipe
(foo@127.0.0.1)2> erl_epmd:names("127.0.0.1").
{ok,[{"riak",59658},{"foo",53341}]}
```

Here's an idea for how to implement it:

``` erlang
nodename_to_atom(Bin) when is_binary(Bin) ->
    try
        binary_to_existing_atom(Bin, utf8)
    catch
        error:badarg ->
            find_node(Bin)
    end.

find_node(Bin) ->
    case re:split(Bin, "@", [{return, list}]) of
        [Name, Host] ->
            case lists:member(Name, get_epmd_names(Host)) of
                true ->
                    binary_to_atom(Bin, utf8);
                false ->
                    erlang:error(badarg, [Bin])
            end;
        _ ->
            erlang:error(badarg, [Bin])
    end.

get_epmd_names(Host) ->
    case erl_epmd:names(Host) of
        {ok, Nodes} ->
            [ Name || {Name, _Port} <- Nodes ];
        _ ->
            []
    end.    
```

cmeiklejohn committed:
Use binary_to_existing_atom.

cmeiklejohn committed:
Be less strict.

cmeiklejohn committed:
Remove another stray log.

cmeiklejohn committed:
Fix spec.

cmeiklejohn committed:
More idiomatic contract.

cmeiklejohn committed:
Needs to be binary to serialize right.

cmeiklejohn committed:
Limit amount of atom conversions.

cmeiklejohn committed:
Dialyzer fix for float typing.

cmeiklejohn committed:
More dialyzer fixes.

jgnewman committed:
removed code that blocks nodes after they are staged

tmcgilchrist committed:
Rebase against jgn-new-ui branch and fix various merge conflicts.

cmeiklejohn committed:
Merge pull request #69 from tmcgilchrist/type_specs

Add type specs.

cmeiklejohn committed:
Fix invalid return value.

cmeiklejohn committed:
Prevent UI glitching during reloads.

jgnewman committed:
the error message now uses a handlebars condition. also fixed some scope issues created by $.ajax. removed some unneeded code from legacy.js

jgnewman committed:
nav now shrinks for small displays

jgnewman committed:
moved some of the code from legacy/gui.js into the views; uncommented tooltip code in gui.js

jgnewman commented:
@cmeiklejohn I moved some stuff from gui.js into the views but I wasn't sure of the best ways to handle everything in there.  I may need your help with what's left as far as doing it the Ember way.

cmeiklejohn commented:
@jgnewman Let's make it more Ember-esque for the next release.  We're close to freeze and I want to ensure we've worked out the usage bugs sooner rather than later.

cmeiklejohn commented:
Closing this PR for a rebased branch.
